### PR TITLE
feat: Userid mapping

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

Added user id mapping feature

## Related issues

-  https://github.com/supertokens/supertokens-node/issues/198

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens/constants.go`
-   ~~[ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)~~
-   [ ] Changes to the version if needed
    -   In `supertokens/constants.go > version variable`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: supertokens/userIdMapping.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+type CreateUserIdMappingResult struct {
+	OK                              *struct{}
+	UnknownSupertokensUserIdError   *struct{}
+	UserIdMappingAlreadyExistsError *struct {
+		DoesSuperTokensUserIdExist bool
+		DoesExternalUserIdExist    bool
+	}
+}
+
+func CreateUserIdMapping(supertokensUserId string, externalUserId string, externalUserIdInfo *string, force bool) (CreateUserIdMappingResult, error) {
```
Comment (rishabhpoddar) [inline supertokens/userIdMapping.go:18]: force bool should be optional with default value as false.

File: supertokens/userIdMapping.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	}
+}
+
+type DeleteUserIdMappingResult struct {
+	OK *struct {
+		DidMappingExist bool
+	}
+}
+
+func DeleteUserIdMapping(userId string, userIdType *string, force bool) (DeleteUserIdMappingResult, error) {
```
Comment (rishabhpoddar) [inline supertokens/userIdMapping.go:130]: force bool should be optional with default value as false.

File: supertokens/userIdMapping.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+type GetUserIdMappingResult struct {
+	OK *struct {
+		SupertokensUserId  string
+		ExternalUserId     string
+		ExternalUserIdInfo *string
+	}
+	UnknownMappingError *struct{}
+}
+
+func GetUserIdMapping(userId string, userIdType *string) (GetUserIdMappingResult, error) {
```
Comment (rishabhpoddar) [inline supertokens/userIdMapping.go:73]: Make userIdType to enum since it only has certain values.

File: supertokens/userIdMapping.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+		}, nil
+	}
+}
+
+func validateUserIdType(userIdType string) error {
+	if userIdType != "SUPERTOKENS" && userIdType != "EXTERNAL" && userIdType != "ANY" {
+		return errors.New("userIdType must be one of SUPERTOKENS, EXTERNAL or ANY")
+	}
+	return nil
+}
```
Comment (rishabhpoddar) [inline supertokens/userIdMapping.go:217]: remove since we are making it an enum

File: supertokens/userIdMapping.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		},
+	}, nil
+}
+
+type UpdateOrDeleteUserIdMappingInfoResult struct {
+	OK                  *struct{}
+	UnknownMappingError *struct{}
+}
+
+func UpdateOrDeleteUserIdMappingInfo(userId string, userIdType *string, externalUserIdInfo *string) (UpdateOrDeleteUserIdMappingInfoResult, error) {
```
Comment (rishabhpoddar) [inline supertokens/userIdMapping.go:171]: need to add a test where you delete the externalUserIdInfo and then call getUserIdMapping and see what happens

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens/userIdMapping.go (hunk 1 above)
Comment (sattvikc) [reply supertokens/userIdMapping.go:18]: Done

File: supertokens/userIdMapping.go (hunk 2 above)
Comment (sattvikc) [reply supertokens/userIdMapping.go:130]: Done

File: supertokens/userIdMapping.go (hunk 3 above)
Comment (sattvikc) [reply supertokens/userIdMapping.go:73]: Done

File: supertokens/userIdMapping.go (hunk 4 above)
Comment (sattvikc) [reply supertokens/userIdMapping.go:217]: Done

File: supertokens/userIdMapping.go (hunk 5 above)
Comment (sattvikc) [reply supertokens/userIdMapping.go:171]: Done

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**